source/common/network/socket_option_factory.cc

```diff
@@ -98,12 +98,19 @@ std::unique_ptr<Socket::Options> SocketOptionFactory::buildLiteralOptions(
                 socket_option.DebugString());
       continue;
     }
+
+    absl::optional<Network::Socket::Type> socket_type = absl::nullopt;
+    if (socket_option.has_type() && socket_option.type().has_stream()) {
```

Comment:
if they're both set, shouldn't we not set socket_type?

Reply:
We are following this new API style to define precedence since `oneof` is now discouraged: https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md

> Prefer multiple fields with defined precedence over boolean overloads of fields or oneof. For example, prefer:

Reply:
oh man I 100% hate that.  Given this isn't a oneof I'd say allowing both makes more sense, especially if we ever add another option like the internal listener type, but I'll let @adisuissa rule on this one and LGTM code as it matches the API.

Let's at least log a warning if both are set since that's at least Envoy style.

Reply:
Done. Logging a warning added when both are set.

Reply:
Yeah, it is unfortunate, and as you said, it was done to allow support for new types, and working with multiple versions of Envoys in a fleet. I guess that if there was API version negotiation, then this could've been avoided by adding the logic on the server side.

I agree about logging, and would also consider adding a counter tracking these misuses in the future.